# NumPy

We always import NumPy using the standard `np` abbreviation.

In [1]:
import numpy as np

## Regular arrays

<iframe width="560" height="315" src="https://www.youtube.com/embed/8Iw8O5qNvvY" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

To learn some different methods for creating NumPy arrays, we will make the following arrays.  (The next one is written as a list, but our primary goal is to make the NumPy array version.)

### Example 1

* [0, 0, 0, ..., 0] (length 10)

Here we use `np.zeros`.  Many Matlab functions have an equivalent version in NumPy, and `zeros` is one such example.

In [2]:
np.zeros(10)

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

Just as an aside, there is also a very elegant way to make a list (not a NumPy array) of 10 zeros using what is called *list comprehension*.  (We will introduce list comprehension more systematically [later in the course](list-comp).)  The underscore `_` in the following is just a way of naming a variable that we will never refer to again; it could be replaced by any variable name and the result would be the same.

Here is the list comprehension way to make a list of 10 zeros.

In [3]:
[0 for _ in range(10)]

[0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

One other way of making this list is based on the following, which uses `+` to concatenate two lists.

In [4]:
[0]+[0]

[0, 0]

So if we want to add `[0]` to itself 10 times, we can use `*`.

In [5]:
[0]*10

[0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

### Example 2

* A five row, seven column matrix containing all 4s.

Here we're trying to make a 5x7 NumPy array of all zeros (then we will add 4 to it).  The following is a common mistake I make.  The error message is not too helpful the first time you see it.

In [6]:
np.zeros(5,7)

TypeError: Cannot interpret '7' as a data type

Notice in the documentation that the first argument for `shape` is supposed to be an integer or a tuple of integers.  The second positional argument is called `dtype`, and NumPy is trying to interpret our second input `7` as a `dtype`.  That was our mistake; we should have given a single input argument as a tuple, not two separate input arguments.

In [7]:
help(np.zeros)

Help on built-in function zeros in module numpy:

zeros(...)
    zeros(shape, dtype=float, order='C', *, like=None)
    
    Return a new array of given shape and type, filled with zeros.
    
    Parameters
    ----------
    shape : int or tuple of ints
        Shape of the new array, e.g., ``(2, 3)`` or ``2``.
    dtype : data-type, optional
        The desired data-type for the array, e.g., `numpy.int8`.  Default is
        `numpy.float64`.
    order : {'C', 'F'}, optional, default: 'C'
        Whether to store multi-dimensional data in row-major
        (C-style) or column-major (Fortran-style) order in
        memory.
    like : array_like, optional
        Reference object to allow the creation of arrays which are not
        NumPy arrays. If an array-like passed in as ``like`` supports
        the ``__array_function__`` protocol, the result will be defined
        by it. In this case, it ensures the creation of an array object
        compatible with that passed in via this arg

Here is the correct way to make a 5x7 NumPy array of all zeros.  Notice the extra set of parentheses around `(5,7)`.  The inner parentheses make a tuple, and the outer parentheses go together with the `zeros` function.

In [8]:
np.zeros((5,7))

array([[0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0.]])

Once we have the array of all zeros, it's easy to get the array of all 4s.  We just add 4, and NumPy automatically "broadcasts" the 4 to all of the entries in the array.

In [9]:
np.zeros((5,7))+4

array([[4., 4., 4., 4., 4., 4., 4.],
       [4., 4., 4., 4., 4., 4., 4.],
       [4., 4., 4., 4., 4., 4., 4.],
       [4., 4., 4., 4., 4., 4., 4.],
       [4., 4., 4., 4., 4., 4., 4.]])

Here is a similar strategy, using the `ones` function instead of the `zeros` function.

In [10]:
np.ones((5,7))

array([[1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.]])

In [11]:
np.ones((5,7))*4

array([[4., 4., 4., 4., 4., 4., 4.],
       [4., 4., 4., 4., 4., 4., 4.],
       [4., 4., 4., 4., 4., 4., 4.],
       [4., 4., 4., 4., 4., 4., 4.],
       [4., 4., 4., 4., 4., 4., 4.]])

If we would rather have integers on the inside (notice how the decimal point goes away), we can specify `int` as the `dtype`.

In [12]:
np.ones((5,7), int)*4

array([[4, 4, 4, 4, 4, 4, 4],
       [4, 4, 4, 4, 4, 4, 4],
       [4, 4, 4, 4, 4, 4, 4],
       [4, 4, 4, 4, 4, 4, 4],
       [4, 4, 4, 4, 4, 4, 4]])

### Example 3

* [0, ..., 100] (length 5, evenly distributed).

In Matlab, this would be made using `linspace`, and NumPy has its own version of `linspace`.

In [13]:
np.linspace(0,100,5)

array([  0.,  25.,  50.,  75., 100.])

(3x3-array)=
### Example 4

* The 3x3 matrix $\begin{pmatrix} 0 & 0 & 0 \\ 1 & 1 & 1 \\ 2 & 2 & 2 \end{pmatrix}$

Here is a first way to make this matrix.  It's again using broadcasting.  For example, when `i` is 2, we broadcast 2 to the second row.  (The second row is specified by `arr[2]`.)  We will later see ways to make this sort of matrix avoiding the for loop.

In [14]:
arr = np.zeros((3,3))
for i in range(3):
    arr[i] = i

In [15]:
arr

array([[0., 0., 0.],
       [1., 1., 1.],
       [2., 2., 2.]])

### Example 5

* The 2x5 matrix $
\begin{pmatrix} 2 & 5 & 8 & 11 & 14 \\ 17 & 20 & 23 & 26 & 29 \end{pmatrix}
$

We start with a nested for loop approach to make this matrix.  We will print out the current version of the array each step through the for loops.  Notice especially how `arr[i,j]` refers to the entry in the i-th row and the j-th column.  The following is almost correct, but it doesn't get the bottom row right.

In [16]:
arr = np.zeros((2,5))
for i in range(2):
    for j in range(5):
        arr[i,j] = 2+3*j
        print(arr)

[[2. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0.]]
[[2. 5. 0. 0. 0.]
 [0. 0. 0. 0. 0.]]
[[2. 5. 8. 0. 0.]
 [0. 0. 0. 0. 0.]]
[[ 2.  5.  8. 11.  0.]
 [ 0.  0.  0.  0.  0.]]
[[ 2.  5.  8. 11. 14.]
 [ 0.  0.  0.  0.  0.]]
[[ 2.  5.  8. 11. 14.]
 [ 2.  0.  0.  0.  0.]]
[[ 2.  5.  8. 11. 14.]
 [ 2.  5.  0.  0.  0.]]
[[ 2.  5.  8. 11. 14.]
 [ 2.  5.  8.  0.  0.]]
[[ 2.  5.  8. 11. 14.]
 [ 2.  5.  8. 11.  0.]]
[[ 2.  5.  8. 11. 14.]
 [ 2.  5.  8. 11. 14.]]


Here is a final version of this approach.  We make the code a little more stable by replacing some of the numbers with variables, for example, we replace the 3 above with the variable `step`.

In [17]:
cols = 5
step = 3
arr = np.zeros((2,cols))
for i in range(2):
    for j in range(cols):
        arr[i,j] = 2+step*j+cols*step*i
print(arr)

[[ 2.  5.  8. 11. 14.]
 [17. 20. 23. 26. 29.]]


Here is an easier way to make the same NumPy array.  Here is a first step: `np.arange` is the NumPy analogue of `range` from Python (or the colon function from Matlab).  (We'll see later that `np.arange` has the advantage over `range` that it can use non-integer step sizes.)

In [18]:
np.arange(2,29,3)

array([ 2,  5,  8, 11, 14, 17, 20, 23, 26])

Remember that the right endpoint is usually not included in Python.  Since we want our array to end at 29, we should list some number bigger than 29 (and less than or equal to 32) as our right endpoint.

In [19]:
np.arange(2,30,3)

array([ 2,  5,  8, 11, 14, 17, 20, 23, 26, 29])

Now we will use the `reshape` method to turn this length-10 one-dimensional NumPy array into a 2x5 two-dimensional NumPy array.

In [20]:
np.arange(2,30,3).reshape((2,5))

array([[ 2,  5,  8, 11, 14],
       [17, 20, 23, 26, 29]])

## Random numbers

<iframe width="560" height="315" src="https://www.youtube.com/embed/UGqkJ2Z3RwY" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

* Make a length 10 NumPy array of random integers between 0 (inclusive) and 40 (exclusive).

You will often see code like the following in code examples, but this is actually the "old" way of getting random numbers in NumPy.

In [2]:
np.random.randint(0, 40, size=10)

array([30, 35, 37,  3,  0, 18, 12, 17, 38, 16])

Notice the portion of the documentation:
> New code should use the ``integers`` method of a ``default_rng()``
        instance instead; please see the :ref:`random-quick-start`
        
For the rest of the course, we will use this `default_rng` method.

In [3]:
help(np.random.randint)

Help on built-in function randint:

randint(...) method of numpy.random.mtrand.RandomState instance
    randint(low, high=None, size=None, dtype=int)
    
    Return random integers from `low` (inclusive) to `high` (exclusive).
    
    Return random integers from the "discrete uniform" distribution of
    the specified dtype in the "half-open" interval [`low`, `high`). If
    `high` is None (the default), then results are from [0, `low`).
    
    .. note::
        New code should use the ``integers`` method of a ``default_rng()``
        instance instead; please see the :ref:`random-quick-start`.
    
    Parameters
    ----------
    low : int or array-like of ints
        Lowest (signed) integers to be drawn from the distribution (unless
        ``high=None``, in which case this parameter is one above the
        *highest* such integer).
    high : int or array-like of ints, optional
        If provided, one above the largest (signed) integer to be drawn
        from the distributi

The `default_rng` approach to random numbers is a good example of object-oriented programming, because we are making a new object, which we call `rng` here, and we are using methods on that object to produce random numbers.

In [5]:
rng = np.random.default_rng()

In [6]:
type(rng)

numpy.random._generator.Generator

The note above said to use the `integers` method, which is what we do here.  This is the *modern* way (as of Summer 2022) to make random numbers in NumPy.

In [8]:
arr = rng.integers(0, 40, size=10)

In [9]:
arr

array([10, 36, 39, 25, 23, 30,  5, 20, 25, 12])

For the rest of this section, we will see a variety of examples of different sorts of methods we can use from our random number generator `rng` object.

* Choose 6 of those numbers (with replacement) and put them into a NumPy array.

In [10]:
rng.choice(arr, size=6)

array([23, 30, 23, 30, 30, 39])

* Make a three row, five column NumPy array of random real numbers between -1 and 4.

The following is what I would have expected, but the `random` method does not allow us to specify a low and a high value for getting random floats.  Instead `random` will always get us random numbers between 0 and 1, and if we want our random numbers in a different range, then we need to rescale the outputs.

In [11]:
rng.random(-1, 4, size=(3,5))

TypeError: random() got multiple values for keyword argument 'size'

Notice that the outputs are between 0 and 1.

In [12]:
help(rng.random)

Help on built-in function random:

random(...) method of numpy.random._generator.Generator instance
    random(size=None, dtype=np.float64, out=None)
    
    Return random floats in the half-open interval [0.0, 1.0).
    
    Results are from the "continuous uniform" distribution over the
    stated interval.  To sample :math:`Unif[a, b), b > a` multiply
    the output of `random` by `(b-a)` and add `a`::
    
      (b - a) * random() + a
    
    Parameters
    ----------
    size : int or tuple of ints, optional
        Output shape.  If the given shape is, e.g., ``(m, n, k)``, then
        ``m * n * k`` samples are drawn.  Default is None, in which case a
        single value is returned.
    dtype : dtype, optional
        Desired dtype of the result, only `float64` and `float32` are supported.
        Byteorder must be native. The default value is np.float64.
    out : ndarray, optional
        Alternative output array in which to place the result. If size is not None,
        it

In [15]:
arr = rng.random((3,5))
arr

array([[0.14373169, 0.04151085, 0.85160258, 0.5625753 , 0.77434184],
       [0.09323801, 0.63691547, 0.83783101, 0.60707909, 0.69066845],
       [0.65647018, 0.45466071, 0.77161217, 0.53020513, 0.71808796]])

The random numbers in `arr` span from 0 to 1 (a distance of 1 apart).  We want to go between -1 and 4, so we eventually want a width of 5.  That is why we will multiply by 5.

In [16]:
arr*5

array([[0.71865844, 0.20755425, 4.25801288, 2.81287651, 3.87170919],
       [0.46619004, 3.18457733, 4.18915507, 3.03539543, 3.45334223],
       [3.28235092, 2.27330357, 3.85806084, 2.65102563, 3.59043979]])

The above array goes between 0 and 5, but we want it to go between -1 and 4, so we subract 1.

In [17]:
arr*5-1

array([[-0.28134156, -0.79244575,  3.25801288,  1.81287651,  2.87170919],
       [-0.53380996,  2.18457733,  3.18915507,  2.03539543,  2.45334223],
       [ 2.28235092,  1.27330357,  2.85806084,  1.65102563,  2.59043979]])

Here is a way to make this array all at once (of course the specific random numbers will be different).

In [18]:
5*rng.random(size=(3,5)) - 1

array([[-0.01973026,  1.61155096,  2.9866675 ,  1.71899874, -0.01933733],
       [ 1.36928355,  2.32786313, -0.00448403, -0.82777187,  0.89141526],
       [ 3.13028253, -0.38060258,  3.05444416,  0.42173001,  1.3847107 ]])

* Make a length 10 NumPy array of random numbers following a normal distribution with mean 2 and standard deviation 0.1.

As just a last example in this section, here is how we can make normally distributed random numbers.

In [19]:
help(rng.normal)

Help on built-in function normal:

normal(...) method of numpy.random._generator.Generator instance
    normal(loc=0.0, scale=1.0, size=None)
    
    Draw random samples from a normal (Gaussian) distribution.
    
    The probability density function of the normal distribution, first
    derived by De Moivre and 200 years later by both Gauss and Laplace
    independently [2]_, is often called the bell curve because of
    its characteristic shape (see the example below).
    
    The normal distributions occurs often in nature.  For example, it
    describes the commonly occurring distribution of samples influenced
    by a large number of tiny, random disturbances, each with its own
    unique distribution [2]_.
    
    Parameters
    ----------
    loc : float or array_like of floats
        Mean ("centre") of the distribution.
    scale : float or array_like of floats
        Standard deviation (spread or "width") of the distribution. Must be
        non-negative.
    size : int o

Notice that the result of the following does seem to be clustered around 2.  That makes sense because we had mean 2 with a relatively low standard deviation (0.1).

In [20]:
rng.normal(2, 0.1, size=10)

array([1.97074003, 2.10991763, 1.98318881, 2.00287628, 2.10707454,
       2.02365598, 1.89343959, 1.97254061, 1.97359416, 1.96238154])

The array is displayed almost like it's two-dimensional, but notice that there is only one set of brackets.  It really is one-dimensional, as we can verify by checking its `shape` attribute.

In [21]:
rng.normal(2, 0.1, size=10).shape

(10,)

## Changing rows and columns

<iframe width="560" height="315" src="https://www.youtube.com/embed/cTBXH47Go_8" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

Here we're going to start with (a slight variant of) the code from an [earlier example](3x3-array).

In [2]:
arr = np.zeros((4,4))
for i in range(4):
    arr[i] = i

In [3]:
arr

array([[0., 0., 0., 0.],
       [1., 1., 1., 1.],
       [2., 2., 2., 2.],
       [3., 3., 3., 3.]])

In [4]:
arr = np.zeros((4,4))

In [6]:
arr

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

If we use indexing of the form `arr[i]` where `arr` is a two-dimensional array, this `arr[i]` will represent the i-th row of `arr`.

In [5]:
arr[2]

array([0., 0., 0., 0.])

Here we set the second row (where as always we start counting from 0) to be all 13s.

In [7]:
arr[2] = 13

In [8]:
arr

array([[ 0.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.],
       [13., 13., 13., 13.],
       [ 0.,  0.,  0.,  0.]])

Accessing a column is slightly harder.  You should read `arr[:,2]` as indicating "every row, the second column".

In [9]:
arr[:,2] = -4

In [10]:
arr

array([[ 0.,  0., -4.,  0.],
       [ 0.,  0., -4.,  0.],
       [13., 13., -4., 13.],
       [ 0.,  0., -4.,  0.]])

The same sort of indexing with a colon could be used with rows.  Here `arr[1,:]` is the same as `arr[1]`.  The `:` in this case can be read as, "every column".  In the following, we go to row 1 and set all of its entries to be 1.

In [11]:
arr[1,:] = 1

In [12]:
arr

array([[ 0.,  0., -4.,  0.],
       [ 1.,  1.,  1.,  1.],
       [13., 13., -4., 13.],
       [ 0.,  0., -4.,  0.]])

In the following, we are storing the second column with the variable name `v`.  This column is a one-dimensional NumPy array.

In [13]:
v = arr[:, 2]

Notice how `v` is displayed horizontally.  Looking at `v`, there is no evidence that this corresponds to a column in our array.

In [14]:
v

array([-4.,  1., -4., -4.])

Here is how a length-1 tuple is displayed in Python, in this case, the length-1 tuple that has only the number 4 in it is written `(4,)`.

In [15]:
v.shape

(4,)

The following code is trying to assign the 3rd row to have the values `[2,10]`.  In this case, we get an error message.  The error message says that we cannot *broadcast* from shape `(2,)` to shape `(4,)`.  In this case, `[2,10]` has shape `(2,)` and `arr[3]` has shape `(4,)`.  Broadcasting is a difficult and important notion in NumPy.

In [16]:
arr[3] = [2,10]

ValueError: could not broadcast input array from shape (2,) into shape (4,)

Here is an example of two different shapes where we can broadcast from one to the other.  In this case, `[1,3,4,7]` has shape `(4,)` and `arr[:]` stands for all the entries in the entire array `arr`, which has shape `(4,4)`.  These two shapes are compatible with respect to broadcasting.

Two shapes in NumPy are compatible with respect to broadcasting if, starting from the right-most dimension (the last dimension), the two dimensions are either equal or one of them is equal to 1.

In [17]:
arr[:] = [1,3,4,7]

In [18]:
arr

array([[1., 3., 4., 7.],
       [1., 3., 4., 7.],
       [1., 3., 4., 7.],
       [1., 3., 4., 7.]])

In [21]:
w = np.array([1,3,4,7])

In [19]:
arr.shape

(4, 4)

In [22]:
w.shape

(4,)

Here is a for-loop version of creating an array `arr`.

In [23]:
arr = np.zeros((3,3))
for i in range(3):
    arr[i] = i

In [24]:
arr

array([[0., 0., 0.],
       [1., 1., 1.],
       [2., 2., 2.]])

Here is a way that uses broadcasting followed by taking the matrix transpose.

In [27]:
arr = np.zeros((3,3))
arr[:] = np.arange(3)
arr.T

array([[0., 0., 0.],
       [1., 1., 1.],
       [2., 2., 2.]])

Here is what `arr` looked like before we took the matrix transpose.  We broadcast from `np.arange(3)` to the 3x3 array `arr`.  If we want to actually change `arr`, we need to reassign `arr` after taking the transpose, using `arr = arr.T`.

In [28]:
arr

array([[0., 1., 2.],
       [0., 1., 2.],
       [0., 1., 2.]])

In [29]:
arr = np.zeros((3,3))
arr[:] = np.arange(3)
arr = arr.T

In [30]:
arr

array([[0., 0., 0.],
       [1., 1., 1.],
       [2., 2., 2.]])

Here is the code for another way (that avoids taking the transpose).  It is a little complicated, so we will break down below what was happening.

In [31]:
arr = np.zeros((3,3))
arr[:] = np.arange(3).reshape((3,1))
arr

array([[0., 0., 0.],
       [1., 1., 1.],
       [2., 2., 2.]])

Let's look in particular at the right-hand side of `arr[:] = np.arange(3).reshape((3,1))`.

In [32]:
x = np.arange(3).reshape((3,1))

In [33]:
x

array([[0],
       [1],
       [2]])

Notice that these dimensions are compatible: for the right-most dimensions, we have 1 and 3, which are compatible (since one of them is equal to 1), and in the next dimension, we have 3 and 3, which are compatible (since they are equal).

In [34]:
x.shape

(3, 1)

In [35]:
arr.shape

(3, 3)

If we tried to reshape using `(2,1)`, we would get an error, because we can't fit 3 numbers into an array of shape `(2,1)`.

In [36]:
y = np.arange(3).reshape((2,1))

ValueError: cannot reshape array of size 3 into shape (2,1)

A very convenient abbreviation to specify "just 1 column" is to use `.reshape((-1,1))`.  Then NumPy is automatically going to replace the `-1` with whatever positive integer is necessary to fill in all the numbers.  In this case, the `-1` gets replaced by `3`.

In [37]:
z = np.arange(3).reshape((-1,1))

In [38]:
z

array([[0],
       [1],
       [2]])

In [39]:
z.shape

(3, 1)

In [40]:
arr

array([[0., 0., 0.],
       [1., 1., 1.],
       [2., 2., 2.]])

As another example of this `-1` syntax, we get the following length-9 one-dimensional NumPy array.  NumPy automatically replaced the `-1` with a `9` in this case, because there were nine numbers in `arr`.

In [41]:
arr.reshape(-1)

array([0., 0., 0., 1., 1., 1., 2., 2., 2.])

## Timing a computation with a NumPy array vs with a list

<iframe width="560" height="315" src="https://www.youtube.com/embed/-rsk3-YRFnY" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

The main goal in this section is to see examples where code written using NumPy can run much faster than code written using base Python.

We'll make some examples using random numbers, so we start out by instantiating a random number generator object.

In [2]:
rng = np.random.default_rng()

Now we make a length one million list of random integers from 1 to 10 (not including 10).  We will be dividing by these numbers, so that's why we don't start with 0.  The `integers` method of `rng` is a NumPy array.

In [3]:
arr = rng.integers(1,10,size=10**6)

Here is the list version of `arr`.

In [4]:
mylist = list(arr)

Here are the first 5 elements in `arr`.

In [5]:
arr[:5]

array([3, 7, 8, 2, 1])

Let's compute `1.5` divided by each of these five numbers.

In [6]:
1.5/arr[:5]

array([0.5       , 0.21428571, 0.1875    , 0.75      , 1.5       ])

If we try to do the same thing with a list, we get an error, saying we can't combine the float 1.5 with the list `mylist` using the operator `/`.

In [7]:
1.5/mylist[:5]

TypeError: unsupported operand type(s) for /: 'float' and 'list'

Here is one way to make the list version, using a for loop.

In [8]:
newlist = []
for x in mylist[:5]:
    newlist.append(1.5/x)

In [9]:
newlist

[0.5, 0.21428571428571427, 0.1875, 0.75, 1.5]

We haven't carefully studied list comprehension yet (we will [next week](list-comp)), but just for contrast, here is how we can make this same list using list comprehension.  The list comprehension is definitely more elegant (often said to be more "Pythonic") than the for loop version.

In [10]:
[1.5/x for x in mylist[:5]]

[0.5, 0.21428571428571427, 0.1875, 0.75, 1.5]

Let's see how the timings compare.

In [11]:
%%timeit
1.5/arr

1.12 ms ± 38.9 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [12]:
%%timeit
newlist = []
for x in mylist:
    newlist.append(1.5/x)

1.66 s ± 35.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [13]:
%%timeit
[1.5/x for x in mylist]

1.62 s ± 27.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


Notice how the NumPy version is over one thousand times faster than the list versions.  My understanding is that there are two important reasons the NumPy version is much faster.  One reason is that the NumPy version of division is *vectorized*, meaning all the divisions are performed at once (no looping is required).  The other reason is that the entries in `arr` are attached to a specific data type, so no time is spent checking the data types of the elements in `arr`.

Also notice that, even though the list comprehension version is more elegant than the for loop version, the timings are very similar.

## Counting using NumPy

<iframe width="560" height="315" src="https://www.youtube.com/embed/7QIKxp2B3iU" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

We've seen examples of making arrays using NumPy and of performing basic arithmetic with NumPy arrays.  In this section, we'll see some examples of counting using NumPy.  For example, if we want to make a probability estimate using the formula "number of successes divided by number of experiments", then being able to count efficiently will be very helpful.

Our example will involve random numbers.  We'll start with a small number (`n = 20`) of random numbers, and then later we will increase `n`.  In general, you should imagine that the bigger the sample, the more accurate our probability estimate should be.

In [2]:
rng = np.random.default_rng()

In [3]:
n = 20

* Make a NumPy array of random integers between -50 and 50 (inclusive).  Also make a list with the same numbers.

In [4]:
arr = rng.integers(-50, 51, size=n)

Here is the list version of the array `arr`.  (Be sure to always use a somewhat random variable name like `mylist`, as opposed to something like `list`, because `list` is already a built-in term in Python.)

In [5]:
mylist = list(arr)

* How many of the integers are strictly less than 10?

Here is a for-loop approach to the above question.

In [6]:
count = 0
for x in mylist:
    if x < 10:
        count += 1

In [7]:
count

13

Here is the list comprehension version of making this count; again, notice how much more concise this is than the for-loop approach.  (We will cover list comprehension more systematically [next week](list-comp)... for now just think of these list comprehension examples as previews.)  We get the same answer `13` as before.

In [8]:
len([x for x in mylist if x < 10])

13

To explain the NumPy approach to counting, it helps to actually look at the values in `arr`.

In [9]:
arr

array([-20,   7,  32, -46,  -2, -16, -12, -39, -22, -38,  17, -20,  49,
        27, -36,   0,  49,   3,  41,  39])

The expression `arr < 10` creates what is called a Boolean array (meaning an array of `True`s and `False`s).  One way to think about this is, it is broadcasting the `< 10` portion to all of the entries in `arr`.  Be sure you understand how the following Boolean array relates to the numbers in `arr`.

In [10]:
arr < 10

array([ True,  True, False,  True,  True,  True,  True,  True,  True,
        True, False,  True, False, False,  True,  True, False,  True,
       False, False])

There are many different possible ways to count `True` in this Boolean array.  I believe the approach which is usually fastest is to use NumPy's `count_nonzero` function.  (Remember that `True` corresponds to `1` and that `False` corresponds to `0`, so in a Boolean array, counting nonzero elements is the same as counting how often `True` occurs.)  Again, we get the same answer `13` as above.

In [11]:
np.count_nonzero(arr < 10)

13

* At what indices do these integers less than 10 occur?

We use the NumPy function `np.nonzero` to find indices where the value is `True` (compare this to `np.count_nonzero` which we found above).  Notice that the resulting output is *not* a NumPy array... the output is a length-1 tuple which contains a NumPy array.

In [12]:
np.nonzero(arr < 10)

(array([ 0,  1,  3,  4,  5,  6,  7,  8,  9, 11, 14, 15, 17]),)

* Make a new array containing only these integers.

We first show the for-loop approach.

In [13]:
newlist = []
for x in mylist:
    if x < 10:
        newlist.append(x)

In [14]:
newlist

[-20, 7, -46, -2, -16, -12, -39, -22, -38, -20, -36, 0, 3]

Here is the list comprehension aproach.  We basically already did this approach above (the only difference is that we included a `len` wrapper above).

In [15]:
[x for x in mylist if x < 10]

[-20, 7, -46, -2, -16, -12, -39, -22, -38, -20, -36, 0, 3]

In [17]:
arr

array([-20,   7,  32, -46,  -2, -16, -12, -39, -22, -38,  17, -20,  49,
        27, -36,   0,  49,   3,  41,  39])

In [16]:
arr < 10

array([ True,  True, False,  True,  True,  True,  True,  True,  True,
        True, False,  True, False, False,  True,  True, False,  True,
       False, False])

The following is what is called "Boolean indexing" or "Applying a Boolean mask".  You should read `arr[arr < 10]` similarly to how you would read `arr[:5]`.  Both are indicating which elements in `arr` to keep.  In the case of `arr[arr < 10]`, the part inside the brackets is a Boolean array, and we keep all the elements where `True` occurs.

In [18]:
arr[arr < 10]

array([-20,   7, -46,  -2, -16, -12, -39, -22, -38, -20, -36,   0,   3])

* If you pick a random integer from such an array, what's the probability it is less than 10?

To get an accurate probability estimate, we want to use as large a value of `n` as can be handled in a reasonably short amount of time; ten million works well in this case.

In [19]:
n = 10**7

In [20]:
arr = rng.integers(-50, 51, size=n)
mylist = list(arr)

Here is the for-loop approach.  We think of `count` as the number of successes and we think of `n` as the number of experiments.

In [21]:
count = 0
for x in mylist:
    if x < 10:
        count += 1
count/n

0.5942684

In [22]:
%%timeit
count = 0
for x in mylist:
    if x < 10:
        count += 1
count/n

930 ms ± 134 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


The list comprehension takes a similar amount of time (it's a little faster than the for-loop version).

In [23]:
%%timeit
len([x for x in mylist if x < 10])

717 ms ± 69.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


Again, the NumPy version is much faster, over 100 times faster than the list comprehension version (and nearly 200 times faster than the for-loop version).

In [24]:
%%timeit
np.count_nonzero(arr < 10)

5.13 ms ± 235 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


To get a sense for how accurate our probability estimate is, we can run the estimate again with new random numbers and see how similar the probability is.

In [25]:
arr = rng.integers(-50, 51, size=n)

In both cases, we get a probability estimate of approximately 0.594.

In [27]:
np.count_nonzero(arr < 10)/n

0.5939019

There are 101 numbers from -50 to 50 (inclusive), and 60 of those are strictly less than 10, so the exact probability is 60/101.  That exact value, 60/101, is very close to our probability estimates, so `n = 10**7` in this case was a big enough sample size to get a very accurate probability estimate.

In [28]:
60/101

0.594059405940594

## Logic in Python and NumPy

<iframe width="560" height="315" src="https://www.youtube.com/embed/vdB2jy5h8Nw" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

General rule of thumb:
* In base Python, use `and`, `or`, `not`
* In NumPy (or pandas), use `&`, `|`, `~` (tilde)

In [2]:
rng = np.random.default_rng()
n = 20

arr = rng.integers(-50, 51, size=n)
mylist = list(arr)

In [3]:
mylist

[-20,
 17,
 -11,
 29,
 -4,
 -50,
 40,
 -36,
 23,
 24,
 -38,
 37,
 6,
 -32,
 -35,
 2,
 16,
 -32,
 -46,
 -47]

* Find all the entries in `arr` and in `mylist` that are strictly between -10 and 10.

Here is the list comprehension solution.  We use `and` because we are doing this in base Python.

In [4]:
[x for x in mylist if (x > -10) and (x < 10)]

[-4, 6, 2]

Here is a more complicated approach, that's just here as an excuse to use `not` and `or`.

In [6]:
[x for x in mylist if not ((x <= -10) or (x >= 10))]

[-4, 6, 2]

Here is the NumPy version.

In [7]:
arr

array([-20,  17, -11,  29,  -4, -50,  40, -36,  23,  24, -38,  37,   6,
       -32, -35,   2,  16, -32, -46, -47])

In [8]:
arr > -10

array([False,  True, False,  True,  True, False,  True, False,  True,
        True, False,  True,  True, False, False,  True,  True, False,
       False, False])

In [9]:
arr < 10

array([ True, False,  True, False,  True,  True, False,  True, False,
       False,  True, False,  True,  True,  True,  True, False,  True,
        True,  True])

If we try to use `and` spelled out between these two Boolean arrays, we will get an error.

In [10]:
(arr > -10) and (arr < 10)

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

We need to use the `&` to combine these Boolean arrays.

In [11]:
(arr > -10) & (arr < 10)

array([False, False, False, False,  True, False, False, False, False,
       False, False, False,  True, False, False,  True, False, False,
       False, False])

Here we find the values using Boolean indexing.

In [12]:
arr[(arr > -10) & (arr < 10)]

array([-4,  6,  2])

Just for fun, here are some other ways to get the same arrays.  We use `~` to negate, and we use the vertical line `|` (sometimes called "pipe") for `or`.

In [13]:
arr[(~(arr <= -10)) & (~(arr >= 10))]

array([-4,  6,  2])

The following is the analogue of the complicated list example from above.  Notice how we replace `not` and `or` in the list version with `~` and `|` in this NumPy version.

In [14]:
arr[~((arr <= -10) | (arr >= 10))]

array([-4,  6,  2])

## The `axis` keyword argument

<iframe width="560" height="315" src="https://www.youtube.com/embed/2FPjPFXBjFs" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

* If you roll four distinct 6-sided dice, what is the probability that the biggest value is 5?

In [2]:
rng = np.random.default_rng()

Here is an example of simulating this experiment using a NumPy random number generator.

In [3]:
rng.integers(1, 7, size=4)

array([4, 3, 5, 1])

Let's try this using ten experiments and a for loop.  (We will eventually want to use many more experiments and we will want to avoid thee for loop.)  Think of `s` as standing for the number of successes.  In the following, we use `np.max` instead of the built-in Python function `max`.  In general, if you have a choice between using a NumPy version or a built-in Python version, you should always prefer the NumPy version.

In [4]:
exps = 10
s = 0

for i in range(exps):
    if np.max(rng.integers(1, 7, size=4)) == 5:
        s += 1

s/exps

0.3

What we just did is very slow.  Even for just one million experiments, the following took 13 seconds.  (It is slow enough that I wanted to use `%%time`, which runs once, instead of `%%timeit`, which runs multiple times and takes an average.)

In [5]:
%%time

exps = 10**6
s = 0

for i in range(exps):
    if np.max(rng.integers(1, 7, size=4)) == 5:
        s += 1

s/exps

CPU times: user 13.7 s, sys: 164 ms, total: 13.9 s
Wall time: 14.3 s


0.284671

The key idea to making this computation more efficient is to work with a two-dimensional array, and to think of each row as corresponding to one experiment.

In [6]:
exps = 10
rng.integers(1, 7, size=(exps,4))

array([[1, 2, 3, 5],
       [5, 6, 1, 4],
       [2, 3, 4, 3],
       [5, 5, 5, 4],
       [4, 5, 4, 3],
       [3, 3, 5, 1],
       [6, 4, 2, 5],
       [3, 2, 2, 4],
       [3, 1, 5, 6],
       [2, 3, 4, 2]])

We can't use `np.max` directly this context (at least, not without using the `axis` keyword argument), because `np.max` will find the maximum of all 40 numbers in this array.

In [7]:
exps = 10
arr = rng.integers(1, 7, size=(exps,4))
print(arr)
np.max(arr)

[[1 4 5 1]
 [6 5 4 4]
 [2 1 3 3]
 [6 6 1 6]
 [6 6 6 6]
 [1 4 2 2]
 [4 1 3 3]
 [5 1 4 5]
 [3 5 3 5]
 [3 4 6 1]]


6

As a first workaround, we could use NumPy's `apply_along_axis` function.  The documentation shows that the first three input arguments are: a function, an axis, and an array.  In our case, the function will be `np.max` and the array will be `arr`.  The main topic of this video is, what is meant by the `axis` keyword argument?

In [8]:
help(np.apply_along_axis)

Help on function apply_along_axis in module numpy:

apply_along_axis(func1d, axis, arr, *args, **kwargs)
    Apply a function to 1-D slices along the given axis.
    
    Execute `func1d(a, *args, **kwargs)` where `func1d` operates on 1-D arrays
    and `a` is a 1-D slice of `arr` along `axis`.
    
    This is equivalent to (but faster than) the following use of `ndindex` and
    `s_`, which sets each of ``ii``, ``jj``, and ``kk`` to a tuple of indices::
    
        Ni, Nk = a.shape[:axis], a.shape[axis+1:]
        for ii in ndindex(Ni):
            for kk in ndindex(Nk):
                f = func1d(arr[ii + s_[:,] + kk])
                Nj = f.shape
                for jj in ndindex(Nj):
                    out[ii + jj + kk] = f[jj]
    
    Equivalently, eliminating the inner loop, this can be expressed as::
    
        Ni, Nk = a.shape[:axis], a.shape[axis+1:]
        for ii in ndindex(Ni):
            for kk in ndindex(Nk):
                out[ii + s_[...,] + kk] = func1d(arr[ii 

In [9]:
arr

array([[1, 4, 5, 1],
       [6, 5, 4, 4],
       [2, 1, 3, 3],
       [6, 6, 1, 6],
       [6, 6, 6, 6],
       [1, 4, 2, 2],
       [4, 1, 3, 3],
       [5, 1, 4, 5],
       [3, 5, 3, 5],
       [3, 4, 6, 1]])

Think of `axis=1` as plugging in one row at a time, and think of `axis=0` as plugging in one column at a time.  In the following example, we get 5 as our first value, because 5 is the maximum of the top row.  We get 6 as our next value, because 6 is the maximum of the next row, and so on.

In [10]:
np.apply_along_axis(np.max, axis=1, arr=arr)

array([5, 6, 3, 6, 6, 4, 4, 5, 5, 6])

What if we used `axis=0` instead?  Then we will get four 6s, because there are four columns, and each column has a 6 as its maximum.

In [11]:
np.apply_along_axis(np.max, axis=0, arr=arr)

array([6, 6, 6, 6])

There is a trick for getting "successes divided by experiments" as our probability estimate in this context.  We make a Boolean array `np.apply_along_axis(np.max, axis=1, arr=arr) == 5`, and then we compute its average.  The intuition is that you can think of the Boolean array as containing 1s and 0s (even though it actually contains Trues and Falses), and if we take its average, that is the same as "number of Trues divided by total number".

In [14]:
exps = 10**6

arr = rng.integers(1, 7, size=(exps,4))
(np.apply_along_axis(np.max, axis=1, arr=arr) == 5).mean()

0.284891

This method is somewhat faster than our for loop method from above, but it still takes about 5 seconds, which is still fairly slow for only having one million experiments. 

In [15]:
%%time

exps = 10**6

arr = rng.integers(1, 7, size=(exps,4))
(np.apply_along_axis(np.max, axis=1, arr=arr) == 5).mean()

CPU times: user 5.27 s, sys: 82.5 ms, total: 5.36 s
Wall time: 5.5 s


0.284142

The problem above is that `np.apply_along_axis` is very general, and is not optimized to the specific maximum computation we're making.  We will replace it by using the `max` method of `arr`.  If we call `max()` on its own, we get the maximum of the whole array, which is not what we want.

In [17]:
exps = 10

arr = rng.integers(1, 7, size=(exps,4))
print(arr)
arr.max()

[[6 6 5 6]
 [5 1 5 5]
 [4 5 4 5]
 [1 3 2 3]
 [2 4 5 6]
 [1 1 1 5]
 [1 2 1 3]
 [2 5 6 1]
 [5 6 1 1]
 [2 1 2 6]]


6

If we check the documentation for `arr.max` (notice that you get documentation for this sort of method by calling `help(arr.max)`, not `help(arr.max()` and not `help(max)`), we see that this `max` method also accepts an `axis` keyword argument, just like `np.apply_along_axis`.

In [18]:
help(arr.max)

Help on built-in function max:

max(...) method of numpy.ndarray instance
    a.max(axis=None, out=None, keepdims=False, initial=<no value>, where=True)
    
    Return the maximum along a given axis.
    
    Refer to `numpy.amax` for full documentation.
    
    See Also
    --------
    numpy.amax : equivalent function



Here is an example of using the `axis` keyword argument on a 10-row, 4-column NumPy array.

In [19]:
exps = 10

arr = rng.integers(1, 7, size=(exps,4))
print(arr)
arr.max(axis=1)

[[5 3 6 4]
 [6 3 1 1]
 [6 6 1 5]
 [1 6 6 2]
 [5 6 4 6]
 [3 1 5 2]
 [1 3 5 1]
 [6 2 4 2]
 [6 1 1 3]
 [2 6 1 5]]


array([6, 6, 6, 6, 6, 5, 5, 6, 6, 6])

In [20]:
exps = 10

arr = rng.integers(1, 7, size=(exps,4))
(arr.max(axis=1) == 5)

array([ True, False, False, False, False, False,  True,  True, False,
        True])

If we try to use `axis=1` as an input argument to `mean` in this case, we will get an error, because the array is one-dimensional, so `axis=1` does not exist.

In [21]:
exps = 10

arr = rng.integers(1, 7, size=(exps,4))
(arr.max(axis=1) == 5).mean(axis=1)

AxisError: axis 1 is out of bounds for array of dimension 1

We could do `axis=0`, but for a one-dimensional NumPy array, we might as well just not provide an `axis` argument.

In [22]:
exps = 10

arr = rng.integers(1, 7, size=(exps,4))
(arr.max(axis=1) == 5).mean(axis=0)

0.4

Here is the best version of our computation.

In [24]:
exps = 10**6

arr = rng.integers(1, 7, size=(exps,4))
(arr.max(axis=1) == 5).mean()

0.285484

The code is not only concise but it is also very efficient.  Whereas even our NumPy code above took over 5 seconds, this is about one hundred times faster.

In [25]:
%%time
exps = 10**6

arr = rng.integers(1, 7, size=(exps,4))
(arr.max(axis=1) == 5).mean()

CPU times: user 46.5 ms, sys: 5.92 ms, total: 52.4 ms
Wall time: 51.3 ms


0.284911

## An exact probability

<iframe width="560" height="315" src="https://www.youtube.com/embed/nS-FIi17grE" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

We ask the same question as last time, but now we will compute the exact probability.

* If you roll four distinct 6-sided dice, what is the probability that the biggest value is 5?

Often computing the exact probability (at least using an exhaustive search where we check every possibility) is not feasible, but in this case, there are only $6^4$ possible values (six different values for each of the four dice), so it's no problem to compute all of the possibilities.

In [2]:
6**4

1296

Here is a correct but inelegant way to make a list of all possible quadruples of integers from 1 to 6 (including 6).

In [3]:
all_dice = []
for i in range(1,7):
    for j in range(1,7):
        for k in range(1,7):
            for l in range(1,7):
                all_dice.append((i,j,k,l))

As a reality check, this list does have the correct length.

In [4]:
len(all_dice)

1296

Here are the first ten values in `all_dice`.  Do you see why it begins this way, and not for example `[(1,1,1,1), (2,1,1,1), ...]`?

In [5]:
all_dice[:10]

[(1, 1, 1, 1),
 (1, 1, 1, 2),
 (1, 1, 1, 3),
 (1, 1, 1, 4),
 (1, 1, 1, 5),
 (1, 1, 1, 6),
 (1, 1, 2, 1),
 (1, 1, 2, 2),
 (1, 1, 2, 3),
 (1, 1, 2, 4)]

What we really want to do is compute the Cartesian product of `[1,2,3,4,5,6]` with itself four times.  The standard way to do that in Python is to use the `itertools` library, which is part of the standard Python libraries (meaning it does not need to be installed).  We won't import the whole module; we'll just import its `product` function.

In [6]:
from itertools import product

In [7]:
help(product)

Help on class product in module itertools:

class product(builtins.object)
 |  product(*iterables, repeat=1) --> product object
 |  
 |  Cartesian product of input iterables.  Equivalent to nested for-loops.
 |  
 |  For example, product(A, B) returns the same as:  ((x,y) for x in A for y in B).
 |  The leftmost iterators are in the outermost for-loop, so the output tuples
 |  cycle in a manner similar to an odometer (with the rightmost element changing
 |  on every iteration).
 |  
 |  To compute the product of an iterable with itself, specify the number
 |  of repetitions with the optional repeat keyword argument. For example,
 |  product(A, repeat=4) means the same as product(A, A, A, A).
 |  
 |  product('ab', range(3)) --> ('a',0) ('a',1) ('a',2) ('b',0) ('b',1) ('b',2)
 |  product((0,1), (0,1), (0,1)) --> (0,0,0) (0,0,1) (0,1,0) (0,1,1) (1,0,0) ...
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /

Here is an example of getting all possible (ordered) pairs of integers from 1 to 6.  It returns an itertools `product` object, but we can convert that object into a list if we want to see all the pairs at once.

In [8]:
product(range(1,7), repeat=2)

In [9]:
list(product(range(1,7), repeat=2))

[(1, 1),
 (1, 2),
 (1, 3),
 (1, 4),
 (1, 5),
 (1, 6),
 (2, 1),
 (2, 2),
 (2, 3),
 (2, 4),
 (2, 5),
 (2, 6),
 (3, 1),
 (3, 2),
 (3, 3),
 (3, 4),
 (3, 5),
 (3, 6),
 (4, 1),
 (4, 2),
 (4, 3),
 (4, 4),
 (4, 5),
 (4, 6),
 (5, 1),
 (5, 2),
 (5, 3),
 (5, 4),
 (5, 5),
 (5, 6),
 (6, 1),
 (6, 2),
 (6, 3),
 (6, 4),
 (6, 5),
 (6, 6)]

Here is another example of using the `product` function without using the `repeat` keyword argument.  In this case, we're getting the Cartesian product of the list `["a", "b", "c"]` with the list `[5, 6]`.  In total, this product has $3 \cdot 2 = 6$ elements.

In [10]:
list(product(["a","b","c"], [5, 6]))

[('a', 5), ('a', 6), ('b', 5), ('b', 6), ('c', 5), ('c', 6)]

One more example.

In [12]:
list(product(["a","b","c"], repeat=3))

[('a', 'a', 'a'),
 ('a', 'a', 'b'),
 ('a', 'a', 'c'),
 ('a', 'b', 'a'),
 ('a', 'b', 'b'),
 ('a', 'b', 'c'),
 ('a', 'c', 'a'),
 ('a', 'c', 'b'),
 ('a', 'c', 'c'),
 ('b', 'a', 'a'),
 ('b', 'a', 'b'),
 ('b', 'a', 'c'),
 ('b', 'b', 'a'),
 ('b', 'b', 'b'),
 ('b', 'b', 'c'),
 ('b', 'c', 'a'),
 ('b', 'c', 'b'),
 ('b', 'c', 'c'),
 ('c', 'a', 'a'),
 ('c', 'a', 'b'),
 ('c', 'a', 'c'),
 ('c', 'b', 'a'),
 ('c', 'b', 'b'),
 ('c', 'b', 'c'),
 ('c', 'c', 'a'),
 ('c', 'c', 'b'),
 ('c', 'c', 'c')]

The above example, using `repeat`, is the exact same as this next example, which explicitly lists the list three times.

In [13]:
list(product(["a","b","c"], ["a","b","c"], ["a","b","c"]))

[('a', 'a', 'a'),
 ('a', 'a', 'b'),
 ('a', 'a', 'c'),
 ('a', 'b', 'a'),
 ('a', 'b', 'b'),
 ('a', 'b', 'c'),
 ('a', 'c', 'a'),
 ('a', 'c', 'b'),
 ('a', 'c', 'c'),
 ('b', 'a', 'a'),
 ('b', 'a', 'b'),
 ('b', 'a', 'c'),
 ('b', 'b', 'a'),
 ('b', 'b', 'b'),
 ('b', 'b', 'c'),
 ('b', 'c', 'a'),
 ('b', 'c', 'b'),
 ('b', 'c', 'c'),
 ('c', 'a', 'a'),
 ('c', 'a', 'b'),
 ('c', 'a', 'c'),
 ('c', 'b', 'a'),
 ('c', 'b', 'b'),
 ('c', 'b', 'c'),
 ('c', 'c', 'a'),
 ('c', 'c', 'b'),
 ('c', 'c', 'c')]

We're now ready to use `product` to simulate every possible roll of four distinct six-sided dice.

In [14]:
temp_list = list(product(range(1,7), repeat=4))

In [15]:
len(temp_list)

1296

Let's convert this to a NumPy array, so we can use some of the earlier methods from this week.

In [16]:
arr = np.array(temp_list)

The array `arr` has 1296 rows and 4 columns.  We think of each row as corresponding to one possible outcome.

In [17]:
arr.shape

(1296, 4)

We want to know, for what proportion of the rows, is the maximum value in that row equal to 5?  Because we want to get the maximum of each row, we use the `axis=1` keyword argument.  (If we wanted to get the maximum of each column, we would instead use `axis=0`.)  We compute the total number of "successes" by using `sum`, and then we divide by the number of rows, which is `arr.shape[0]`.

In [19]:
(arr.max(axis=1) == 5).sum()/arr.shape[0]

0.2847222222222222

An abbreviation is to apply the `mean` method to our Boolean array.

In [20]:
(arr.max(axis=1) == 5).mean()

0.2847222222222222

Just for fun, here is another computation of the exact probability.  The first number in the difference, $(5/6)^4$, is equal to the probability that none of the dice lands 6, and then we are subtracting the probability that none of the dice lands 5 or 6.

In [21]:
(5/6)**4-(4/6)**4

0.2847222222222223

## `any` and `all`

<iframe width="560" height="315" src="https://www.youtube.com/embed/0GeviCLROaw" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

Make a 100 row, 4 column NumPy array `arr` of random real numbers between 0 and 1.

In [2]:
rng = np.random.default_rng()

In [4]:
arr = rng.random(size=(100,4))

* Find the sub-array of `arr` containing all rows in which at least one number is bigger than 0.9.

Here are the first 5 rows of `arr`.

In [6]:
arr[:5]

array([[0.02679582, 0.39964722, 0.38608334, 0.9557704 ],
       [0.62903247, 0.17981028, 0.48643802, 0.16059557],
       [0.49612872, 0.34270062, 0.00862007, 0.77873829],
       [0.55550408, 0.37485681, 0.20417908, 0.82782464],
       [0.51663663, 0.85479878, 0.48625127, 0.88475327]])

Here is a Boolean array corresponding to which elements are strictly greater than 5.

In [7]:
arr[:5] > 0.9

array([[False, False, False,  True],
       [False, False, False, False],
       [False, False, False, False],
       [False, False, False, False],
       [False, False, False, False]])

For each row, we want to know if any value in that row is `True`.  To compute this, we will use the `any` method, together with `axis=1`.  The reason we're using `axis=1` is because we are plugging in a full row at a time.

In [8]:
(arr[:5] > 0.9).any(axis=1)

array([ True, False, False, False, False])

If we wanted to count how many of these first 5 rows contain a number bigger than 0.9, we could use `sum`.

In [9]:
(arr[:5] > 0.9).any(axis=1).sum()

1

The question asked for us to find the rows which contain a number bigger than 0.9, so we will use Boolean indexing.  We are plugging in the `array([ True, False, False, False, False])` from above into the square brackets, which is like saying, keep the top row, discard the next four rows.

In [10]:
arr[:5][(arr[:5] > 0.9).any(axis=1)]

array([[0.02679582, 0.39964722, 0.38608334, 0.9557704 ]])

We now do the same thing to the full array.  This will be a little easier to read, because we can get rid of the `[:5]` slices.

In [11]:
arr[(arr > 0.9).any(axis=1)]

array([[0.02679582, 0.39964722, 0.38608334, 0.9557704 ],
       [0.89128322, 0.97335335, 0.1821295 , 0.52423414],
       [0.4629412 , 0.71844233, 0.91362938, 0.04770196],
       [0.61030793, 0.15565872, 0.15616296, 0.9572168 ],
       [0.45523301, 0.85912337, 0.34660356, 0.91882733],
       [0.14055923, 0.86443333, 0.95413845, 0.94380163],
       [0.07850217, 0.96370185, 0.21801727, 0.22118187],
       [0.10730651, 0.56976149, 0.95367251, 0.7782287 ],
       [0.95705318, 0.11012904, 0.1380347 , 0.67057015],
       [0.95930433, 0.06501316, 0.67658178, 0.20159313],
       [0.2457492 , 0.67240127, 0.94358665, 0.90993476],
       [0.37975889, 0.92498652, 0.99327127, 0.7283782 ],
       [0.37093209, 0.45237422, 0.91160131, 0.96612184],
       [0.52254037, 0.19969189, 0.94795525, 0.15420271],
       [0.90339743, 0.59821255, 0.62045249, 0.54939576],
       [0.48895101, 0.52105202, 0.96297772, 0.45230437],
       [0.35246537, 0.32415082, 0.86599985, 0.98355888],
       [0.32443071, 0.94130683,

* Find the sub-array of `arr` containing all rows in which no numbers are between 0.4 and 0.6.

One way to think about this is, we want the rows in which every element in that row is less than 0.4 or bigger than 0.6.  (We allow a single row to have both smaller and bigger elements; all we care about is that it does not contain any values that start with 0.4 or 0.5.)  Since we want a condition to be true for all the numbers in the row, we use `all(axis=1)`.

In [12]:
((arr < 0.4) | (arr > 0.6)).all(axis=1)

array([ True, False, False, False, False, False, False, False, False,
       False, False,  True, False,  True,  True, False,  True,  True,
       False, False, False,  True,  True,  True,  True,  True,  True,
        True,  True, False, False, False, False, False, False, False,
       False, False,  True, False, False,  True, False,  True,  True,
       False, False, False, False, False,  True,  True,  True, False,
        True,  True, False,  True, False, False, False, False, False,
        True,  True,  True, False, False, False,  True, False, False,
        True, False,  True,  True, False, False, False, False, False,
        True, False, False, False, False, False, False,  True, False,
        True, False, False, False,  True,  True, False, False,  True,
       False])

As a reality check, our Boolean array here is length 100.

In [13]:
len(((arr < 0.4) | (arr > 0.6)).all(axis=1))

100

Here again we're using Boolean indexing.  We're keeping all the rows of `arr` which satisfy this condition.  As you scroll through this array, notice how all the values that appear are either less than 0.4 or bigger than 0.6.

In [14]:
arr[((arr < 0.4) | (arr > 0.6)).all(axis=1)]

array([[0.02679582, 0.39964722, 0.38608334, 0.9557704 ],
       [0.61030793, 0.15565872, 0.15616296, 0.9572168 ],
       [0.64463239, 0.62136197, 0.83285561, 0.69102624],
       [0.19806438, 0.28789025, 0.16912325, 0.71827727],
       [0.14055923, 0.86443333, 0.95413845, 0.94380163],
       [0.07850217, 0.96370185, 0.21801727, 0.22118187],
       [0.95705318, 0.11012904, 0.1380347 , 0.67057015],
       [0.6703015 , 0.81869015, 0.31666259, 0.67068789],
       [0.74644917, 0.06370154, 0.34252816, 0.06391344],
       [0.06099061, 0.08149432, 0.29021873, 0.85693158],
       [0.80698662, 0.31211537, 0.68117698, 0.78836747],
       [0.95930433, 0.06501316, 0.67658178, 0.20159313],
       [0.2457492 , 0.67240127, 0.94358665, 0.90993476],
       [0.37975889, 0.92498652, 0.99327127, 0.7283782 ],
       [0.35246537, 0.32415082, 0.86599985, 0.98355888],
       [0.32443071, 0.94130683, 0.86154932, 0.95131523],
       [0.75210216, 0.38678934, 0.19343926, 0.65172332],
       [0.75785393, 0.97896958,

## From binary to decimal, version 1

<iframe width="560" height="315" src="https://www.youtube.com/embed/bXyqvBQ4wP4" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

$$
\begin{pmatrix} 1 & 0 & 1 & 0 \\ 0 & 1 & 0 & 0 \\ 1 & 1 & 1 & 1 \end{pmatrix} \mapsto \begin{pmatrix} 10 \\ 4 \\ 15 \end{pmatrix}
$$

* Write a function `to_bin` which takes in an m-by-n NumPy array of 0s and 1s, and as output returns a length m NumPy array of the corresponding integers, where we think of each row as representing the binary digits of an integer.

For example, we think of the top row as representing $1 \cdot 2^3 + 0 \cdot 2^2 + 1 \cdot 2^1 + 0 \cdot 2^0 = 10$.

(If I were writing this again, I would call the function `from_bin` instead of `to_bin`, but I'll stick with `to_bin` to match what's in the video.)

Let's start by getting a random array that can be our test input.

In [2]:
rng = np.random.default_rng()

In [3]:
arr = rng.integers(2, size=(10,4))

We assign `m` to be the number of rows and `n` to be the number of columns, using what is called tuple unpacking.

In [4]:
m,n = arr.shape

In [5]:
m

10

In [6]:
n

4

We want to multiply the left-most column by $2^3$, the next column by $2^2$, and so on.  Here are the exponents.  The middle `-1` is telling NumPy where to stop (stop before -1), and the right-most `-1` is telling NumPy what the step size should be (go down by 1 each step).

In [7]:
np.arange(n-1, -1, -1)

array([3, 2, 1, 0])

Here are the coefficients we want to multiply the columns by.

In [8]:
2**np.arange(n-1, -1, -1)

array([8, 4, 2, 1])

Using broadcasting, it's very easy to make the multiplication.  And to add up along the rows, we use `sum(axis=1)`.

In [12]:
(arr*(2**np.arange(n-1, -1, -1))).sum(axis=1)

array([12,  7, 11, 15,  4,  1,  0, 12, 14,  9])

In [13]:
def to_bin(arr):
    _,n = arr.shape
    return (arr*(2**np.arange(n-1, -1, -1))).sum(axis=1)

In [14]:
to_bin(arr)

array([12,  7, 11, 15,  4,  1,  0, 12, 14,  9])

Here we check on a smaller array.

In [15]:
to_bin(np.array([[1,0],[1,1]]))

array([2, 3])

## From binary to decimal, version 2

<iframe width="560" height="315" src="https://www.youtube.com/embed/AzN2i8NaSRY" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

$$
\begin{pmatrix} 1 & 0 & 1 & 0 \\ 0 & 1 & 0 & 0 \\ 1 & 1 & 1 & 1 \end{pmatrix} \mapsto \begin{pmatrix} 10 \\ 4 \\ 15 \end{pmatrix}
$$

* Write a new function, `to_bin2`, which takes in an m-by-n NumPy array of 0s and 1s, and as output returns a length m NumPy array of the corresponding integers, where we think of each row as representing the binary digits of an integer.

In [2]:
rng = np.random.default_rng()
arr = rng.integers(2, size=(10,4))

In [3]:
arr

array([[0, 1, 1, 1],
       [0, 0, 1, 0],
       [1, 0, 0, 1],
       [1, 0, 1, 1],
       [0, 1, 1, 1],
       [1, 0, 0, 1],
       [0, 1, 0, 1],
       [0, 0, 1, 0],
       [1, 1, 0, 0],
       [0, 0, 1, 0]])

Here is the "usual" way of converting from a string to an integer.

In [4]:
int("0111")

111

But we can also tell Python that the string should be thought of as being written in binary.  For example, 0111 in binary should be thought of as $4+2+1 = 7$.  In the following, we're specifying that the number is being written in "base-2".

In [5]:
int("0111", 2)

7

Our "usual" number system is written in base-10.

In [6]:
int("0111", 10)

111

Here we make a length-4 NumPy array.

In [7]:
z = np.array([0,1,1,1])

Or it's probably more elegant to just take the top row of `arr`.

In [8]:
z = arr[0]

In [9]:
z

array([0, 1, 1, 1])

How can we convert that into the string "0111"?  I feel like there might be an easier way that I'm missing, but this is what I've come up with.  As a preliminary step, here is a reminder of list comprehension which we won't officially cover in Math 9 until [next week](list-comp).  In the following, we use list comprehension to put the integers 0, 1, 1, 1 into a list.

In [10]:
[x for x in z]

[0, 1, 1, 1]

Here we use list comprehension to put the strings "0", "1", "1", "1" into a list.  It is the same as the previous command, except that we convert each integer to a string using `str`.

In [11]:
[str(x) for x in z]

['0', '1', '1', '1']

I had initially thought the following would concatenate these strings (which would make sense, because `"0"+"1"` concatenates the strings to produce the string `"01"`), but it doesn't work.

In [12]:
sum([str(x) for x in z])

TypeError: unsupported operand type(s) for +: 'int' and 'str'

The following is the suggested way in Python to concatenate all the strings in a list together.

In [13]:
''.join([str(x) for x in z])

'0111'

We want to apply that idea to every row in `arr`.  Let's define a helper function to use.  Later we will see a way to write this sort of short function in a more concise way, using "lambda functions".  (Caution.  A common programming mistake is to use something like `print` instead of `return`.  That's very incorrect.  If you want your function to have an output, you need to use `return`.)

In [14]:
def helper(z):
    return ''.join([str(x) for x in z])

Let's test out our `helper` function on the 3rd row of `arr`.

In [15]:
helper(arr[3])

'1011'

In [16]:
arr[3]

array([1, 0, 1, 1])

In [17]:
arr

array([[0, 1, 1, 1],
       [0, 0, 1, 0],
       [1, 0, 0, 1],
       [1, 0, 1, 1],
       [0, 1, 1, 1],
       [1, 0, 0, 1],
       [0, 1, 0, 1],
       [0, 0, 1, 0],
       [1, 1, 0, 0],
       [0, 0, 1, 0]])

Here is a version `helper2` that treats the row as if it corresponds to binary digits.  We could do this all in one line, but it is more readable to break it up into two steps.

In [18]:
def helper2(z):
    temp = ''.join([str(x) for x in z])
    return int(temp, 2)

In [19]:
helper2(arr[3])

11

Here we're using `apply_along_axis` again.  The function we want to apply is `helper2`, and we are applying along rows, so we use `axis=1`.

In [20]:
def to_bin2(A):
    return np.apply_along_axis(helper2, axis=1, arr=A)

This doesn't work because `A` is not defined.

In [21]:
to_bin2(A)

NameError: name 'A' is not defined

In [22]:
arr

array([[0, 1, 1, 1],
       [0, 0, 1, 0],
       [1, 0, 0, 1],
       [1, 0, 1, 1],
       [0, 1, 1, 1],
       [1, 0, 0, 1],
       [0, 1, 0, 1],
       [0, 0, 1, 0],
       [1, 1, 0, 0],
       [0, 0, 1, 0]])

Notice how, for example, the top row corresponds to 7, the next row corresponds to 2, and so on.

In [23]:
to_bin2(arr)

array([ 7,  2,  9, 11,  7,  9,  5,  2, 12,  2])

## Raising errors

<iframe width="560" height="315" src="https://www.youtube.com/embed/1F9A0K7A6ak" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

* Edit the `to_bin` function so that it raises an error if the input is not a NumPy array and it also raises an error if an entry is not 0 or 1.

In [2]:
rng = np.random.default_rng()
arr = rng.integers(2, size=(10,4))

In [4]:
arr

array([[1, 1, 1, 1],
       [0, 0, 1, 1],
       [1, 0, 1, 1],
       [1, 1, 1, 0],
       [0, 0, 1, 1],
       [0, 1, 0, 1],
       [1, 1, 0, 1],
       [1, 1, 1, 1],
       [1, 1, 0, 0],
       [0, 0, 1, 1]])

In [5]:
type(arr)

numpy.ndarray

Here is the usual way to check the `type` of a variable, using the function `isinstance`.  It doesn't literally check that the type is exactly a NumPy array, but what this checks is usually good enough.  (Technically, it checks if `arr` is a NumPy array or is a subclass of the class NumPy array.)

In [6]:
isinstance(arr, np.ndarray)

True

In [7]:
isinstance(arr, list)

False

Here is how we raise an error in Python.  In this case, we will raise a TypeError.  Don't worry about learning the many different error types in Python; for Math 9, the important thing is to know the general syntax of how to raise an error.

In [8]:
def to_bin(arr):
    if not isinstance(arr, np.ndarray):
        raise TypeError("Input should be a NumPy array")
    _,n = arr.shape
    return (arr*(2**np.arange(n-1, -1, -1))).sum(axis=1)

In [9]:
to_bin(arr)

array([15,  3, 11, 14,  3,  5, 13, 15, 12,  3])

Notice how the error message we wrote, "Input should be a NumPy array", shows up in the following error message.

In [10]:
to_bin(list(arr))

TypeError: Input should be a NumPy array

Now we will work on the second portion, namely, checking that all of the entries are 0 or 1.

In [11]:
arr == 0

array([[False, False, False, False],
       [ True,  True, False, False],
       [False,  True, False, False],
       [False, False, False,  True],
       [ True,  True, False, False],
       [ True, False,  True, False],
       [False, False,  True, False],
       [False, False, False, False],
       [False, False,  True,  True],
       [ True,  True, False, False]])

In [12]:
arr == 1

array([[ True,  True,  True,  True],
       [False, False,  True,  True],
       [ True, False,  True,  True],
       [ True,  True,  True, False],
       [False, False,  True,  True],
       [False,  True, False,  True],
       [ True,  True, False,  True],
       [ True,  True,  True,  True],
       [ True,  True, False, False],
       [False, False,  True,  True]])

The array is acceptable if each entry is zero or one, so we use the vertical line `|` for "or".

In [13]:
(arr == 0) | (arr == 1)

array([[ True,  True,  True,  True],
       [ True,  True,  True,  True],
       [ True,  True,  True,  True],
       [ True,  True,  True,  True],
       [ True,  True,  True,  True],
       [ True,  True,  True,  True],
       [ True,  True,  True,  True],
       [ True,  True,  True,  True],
       [ True,  True,  True,  True],
       [ True,  True,  True,  True]])

Because we are checking each entry (not each entry in a given row), we use the `all` method with no `axis` argument.

In [14]:
((arr == 0) | (arr == 1)).all()

True

In [15]:
def to_bin(arr):
    if not isinstance(arr, np.ndarray):
        raise TypeError("Input should be a NumPy array")
    if not ((arr == 0) | (arr == 1)).all():
        raise ValueError("All entries should be 0 or 1")
    _,n = arr.shape
    return (arr*(2**np.arange(n-1, -1, -1))).sum(axis=1)

In [16]:
to_bin(arr)

array([15,  3, 11, 14,  3,  5, 13, 15, 12,  3])

In [17]:
arr2 = arr

We haven't emphasized the following type of assignment much, but here is how you set the 5th row, 2nd column to be 6.

In [18]:
arr2[5,2] = 6

In [19]:
arr2

array([[1, 1, 1, 1],
       [0, 0, 1, 1],
       [1, 0, 1, 1],
       [1, 1, 1, 0],
       [0, 0, 1, 1],
       [0, 1, 6, 1],
       [1, 1, 0, 1],
       [1, 1, 1, 1],
       [1, 1, 0, 0],
       [0, 0, 1, 1]])

Here this raises the error we chose, because one of the entries is 6.

In [20]:
to_bin(arr2)

ValueError: All entries should be 0 or 1

A big "gotcha" in Python is that our assignment above also changed the original `arr` array.  (NumPy does this to save memory space.)

In [21]:
to_bin(arr)

ValueError: All entries should be 0 or 1

Notice how `arr` got changed, even though we never changed it ourselves, we only changed `arr2`.

In [22]:
arr

array([[1, 1, 1, 1],
       [0, 0, 1, 1],
       [1, 0, 1, 1],
       [1, 1, 1, 0],
       [0, 0, 1, 1],
       [0, 1, 6, 1],
       [1, 1, 0, 1],
       [1, 1, 1, 1],
       [1, 1, 0, 0],
       [0, 0, 1, 1]])

If you want to make sure the original array doesn't get changed, use the `copy` method.

In [23]:
arr3 = arr.copy()

In [24]:
arr3

array([[1, 1, 1, 1],
       [0, 0, 1, 1],
       [1, 0, 1, 1],
       [1, 1, 1, 0],
       [0, 0, 1, 1],
       [0, 1, 6, 1],
       [1, 1, 0, 1],
       [1, 1, 1, 1],
       [1, 1, 0, 0],
       [0, 0, 1, 1]])

Here we set all the even-indexed columns in row 6 to be equal to 99.

In [25]:
arr3[6,::2] = 99

In [26]:
arr3

array([[ 1,  1,  1,  1],
       [ 0,  0,  1,  1],
       [ 1,  0,  1,  1],
       [ 1,  1,  1,  0],
       [ 0,  0,  1,  1],
       [ 0,  1,  6,  1],
       [99,  1, 99,  1],
       [ 1,  1,  1,  1],
       [ 1,  1,  0,  0],
       [ 0,  0,  1,  1]])

Because we used the `copy` method when we created `arr3`, the original array `arr` did not get changed by our assignment.

In [27]:
arr

array([[1, 1, 1, 1],
       [0, 0, 1, 1],
       [1, 0, 1, 1],
       [1, 1, 1, 0],
       [0, 0, 1, 1],
       [0, 1, 6, 1],
       [1, 1, 0, 1],
       [1, 1, 1, 1],
       [1, 1, 0, 0],
       [0, 0, 1, 1]])